core/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/core/model/HttpInput.kt

```diff
@@ -0,0 +1,153 @@
+    }
+
+    override fun toXContent(builder: XContentBuilder, params: ToXContent.Params): XContentBuilder {
+        return builder.startObject()
+                .startObject(HTTP_FIELD)
+                .field(SCHEME_FIELD, scheme)
+                .field(HOST_FIELD, host)
+                .field(PORT_FIELD, port)
+                .field(PATH_FIELD, path)
+                .field(PARAMS_FIELD, this.params)
+                .field(URL_FIELD, url)
+                .field(CONNECTION_TIMEOUT_FIELD, connection_timeout)
+                .field(SOCKET_TIMEOUT_FIELD, socket_timeout)
+                .endObject()
+                .endObject()
+    }
+
+    override fun name(): String {
+        return HTTP_FIELD
+    }
+
+    companion object {
+        const val SCHEME_FIELD = "scheme"
+        const val HOST_FIELD = "host"
+        const val PORT_FIELD = "port"
+        const val PATH_FIELD = "path"
+        const val PARAMS_FIELD = "query_params"
+        const val URL_FIELD = "url"
+        const val CONNECTION_TIMEOUT_FIELD = "connection_timeout"
+        const val SOCKET_TIMEOUT_FIELD = "socket_timeout"
+        const val HTTP_FIELD = "http"
+
+        val XCONTENT_REGISTRY = NamedXContentRegistry.Entry(Input::class.java, ParseField("http"), CheckedFunction { parseInner(it) })
+
+        /**
+         * This parse function uses [XContentParser] to parse JSON input and store corresponding fields to create a [HttpInput] object
+         */
+        @JvmStatic @Throws(IOException::class)
+        private fun parseInner(xcp: XContentParser): HttpInput {
+            var scheme = "http"
+            var host = ""
+            var port: Int = -1
+            var path = ""
+            var params: Map<String, String> = mutableMapOf()
+            var url = ""
+            var connectionTimeout = 5
+            var socketTimeout = 10
+
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, xcp.currentToken(), xcp::getTokenLocation)
+
+            while (xcp.nextToken() != XContentParser.Token.END_OBJECT) {
+                val fieldName = xcp.currentName()
+                xcp.nextToken()
+                when (fieldName) {
+                    SCHEME_FIELD -> scheme = xcp.text()
+                    HOST_FIELD -> host = xcp.text()
+                    PORT_FIELD -> port = xcp.intValue()
+                    PATH_FIELD -> path = xcp.text()
+                    PARAMS_FIELD -> params = xcp.mapStrings()
+                    URL_FIELD -> url = xcp.text()
+                    CONNECTION_TIMEOUT_FIELD -> connectionTimeout = xcp.intValue()
+                    SOCKET_TIMEOUT_FIELD -> socketTimeout = xcp.intValue()
+                }
+            }
+            return HttpInput(scheme, host, port, path, params, url, connectionTimeout, socketTimeout)
+        }
+    }
+
+    /**
+     * Helper function to check whether one of url or scheme+host+port+path+params is defined.
+     */
+    private fun validateFields(): Boolean {
```

Comment:
Going off of its name, it look like `validateFields()` should return `true` when the fields are valid. In this case it might be more accurate to call this `areFieldsValid()`.

Also, I might be mistaken but it looks the boolean checks are inverted. `if (url.isNotEmpty())` will only check the `host/port/path/params` when the `url` is not empty, where it should only be checking those values if a `url` is not given. The current logic will return `true` by default when `url` is empty which does not validate the `host/port/path/params`.

Perhaps something like this would make more sense:
```
if (url.isEmpty()) {
    return !(host.isEmpty() && (port == -1) && path.isEmpty() && params.isEmpty())
}

return true
```

Reply:
Thanks @qreshi for giving the detailed comments! 👍 I will look at the codes.